src/main/java/io/vertx/codegen/TypeValidator.java

```diff
@@ -111,6 +116,16 @@ private static boolean isValidNonCallableType(Element elem, TypeInfo type, boole
     return false;
   }
 
+  private static boolean isValidKotlinCompanionObjet(Element elem, TypeInfo type) {
+    String qualifiedName = type.getName(); // e.g. io.vertx.test.codegen.testapi.kotlin.InterfaceWithCompanionObject.Companion
+    String simpleName = type.getSimpleName(); // e.g. Companion
+    String enclosingElementName = elem.getEnclosingElement().getSimpleName().toString(); // e.g. InterfaceWithCompanionObject
```

Comment:
can you try instead to look at the ElementKind of the enclosing element and return true when this element is `ElementKind.CLASS` ?

Reply:
My intention was to limit this to Kotlin companion objects -- where the type `Companion` is defined in the same class  -- as I didn't know if other scenarios were valid.

Will replacing this with a simple `ElementKind` check cause issues?

**That is, both of these are valid:**
https://github.com/vert-x3/vertx-codegen/blob/d14c30d5fccfed8c51b020066d4e014891d47564/src/test/java/io/vertx/test/codegen/testapi/kotlin/InterfaceWithCompanionObject.java#L10-L26

https://github.com/vert-x3/vertx-codegen/blob/d14c30d5fccfed8c51b020066d4e014891d47564/src/test/java/io/vertx/test/codegen/testapi/kotlin/InterfaceWithUnrelatedCompanionClass.java#L9-L11

Reply:
what is the exact precise condition that isValidKotlinCompanionObjet needs to meet to return true with respect with to Kotlin ?

Reply:
ping

Reply:
> what is the exact precise condition that isValidKotlinCompanionObjet needs to meet to return true with respect with Kotlin companion objeect ?

Apologies, missed your response.

---

**Context**

Given the following Kotlin implementation:
```kotlin
@ProxyGen
interface SomeDatabaseService {

    fun save(item: Any, handler: Handler<AsyncResult<Void>>)

    @GenIgnore
    companion object {
        @JvmStatic
        fun create(vertx: Vertx): SomeDatabaseService = SomeDatabaseServiceImpl(vertx)
    }
}
```

The resulting Java code is (roughly):
```java
@ProxyGen
public interface SomeDatabaseService {

  SomeDatabaseService.Companion Companion = SomeDatabaseService.Companion.$$INSTANCE;

  void save (Object var1, Handler var3);

  static void create(Vertx vertx) { Companion.create(vertx); }

  @GenIgnore
  public static final class Companion {
    private Companion() {
    }

    static final SomeDatabaseService.Companion $$INSTANCE;

    @JvmStatic
    public final SomeDatabaseService create(Vertx vertx) {
      return new  SomeDatabaseServiceImpl(vertx);
    }

    static {
      SomeDatabaseService.Companion var0 = new SomeDatabaseService.Companion();
      $$INSTANCE = var0;
    }
  }
}
```

Kotlin generates a `Companion` class and `Companion` field in `SomeDatabaseService`; however, the `@GenIgnore` annotation is only applied to the class. (If the companion object is named `Foo` the class and field will instead be `Foo`)

**`isValidKotlinCompanionObject` Condition**

The method is meant to determine if a field is referencing an inner Companion class. The easiest approach I could think of is testing if the field type's qualified name (`io.vertx.example.SomeDatabaseService.Companion`) equals the enclosing class's qualified name + the field's name (`io.vertx.SomeDatabaseService` + `Companion`).